eclipse-theia/theia
# Support SSH remote development feature

#### What it does

Closes https://github.com/eclipse-theia/theia/issues/12390

This change adds the functionality to connect to remote backends using SSH. It adds infrastructure to support other types of connections as well (i.e. WSL or Devcontainers), but implements the functionality only for SSH connections.

Note that this change supports only the Electron version of Theia.

#### How to test

Requires an SSH capable system to which the connection can be established. Note that due to https://github.com/eclipse-theia/theia/issues/12782, this requires a fairly updated system with `glibc >= 2.33` (on Linux at least). Also due to a limitation on GitHub actions, it requires a x64 machine. I've simply used a Gitpod workspace.

1. Start the application and run the `SSH: Connect to Host...` command (or click on the remote status bar entry in the bottom left of the application). Enter the remote system address.
2. The local system should connect to the remote system successfully (or display a message on failure).
3. The remote system can be used as expected, all functionality works as it should.

#### Follow-Ups

* Add more target architectures. Currently, only x64 machines are supported by the feature, meaning no arm64 (and therefore no Mac M1 or M2).
* The backend doesn't copy the plugins from the original user to the remote host yet. This will come in a separate PR.
* The security constraints for electron token validation have been slightly relaxed. I plan to reimplement that feature as a more generalized security token for both browser and electron.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
